### DISCUSSION TITLE
2021-09-06 (future automatic release)

### DISCUSSION THREAD

**Discussion by pipcet** [pipcet/pearl#19] [Announcements]
If all goes well, I've timed things so that the new release will be built and available shortly after midnight UTC, so it will be at https://github.com/pipcet/pearl/releases/tag/2021-09-06 .

I'm writing these optimistic instructions under the assumption that the release does indeed work the way my test builds did. I hope that's not too much hubris.

This release adds a new way of bootstrapping pearl: pearl.pl, a perl script that runs in the (very limited) Recovery Mode environment and automates installation to a certain extent. This finally makes things work out of the box on my MBP.

To install pearl, boot into Recovery Mode, select Options, then Terminal from (which?) menu, then enter:

```
curl https://github.com/pipcet/pearl/releases/latest/download/pearl.pl|perl
```

Please make sure that you're using HTTPS, as you're feeding a perl script directly into a perl interpreter running as the root user. (It would be a good idea to verify the integrity of the file with `sha512sum`, but that's not available in the Recovery Mode environment. `md5` is, but that's no longer a secure choice.)

After the download completes, make sure you understand what you are doing and, if you decide to go ahead, enter your username and password. This is as good a time as any to remind you there are no warranties and you're doing this at your own risk.

Ideally, the following should happen:

The system reboots automatically after a while. The screen goes dark for longer than you expect. Do nothing and wait.

The screen comes back on with a full-screen text console and some text scrolls by. There's a one-second delay after the model number is printed, and it might be a good idea to write that down, though it can also be determined lateron. Do nothing and wait.

The screen will come on again with four terminal windows: the upper left is the menu for interaction, the upper right is a general help/system status window, the lower left is for kernel messages, the lower right is for a shell prompt. There's a heartbeat and a countdown to when Debian will boot automatically. Do nothing and wait, or, if you're impatient and have the right keyboard, hit Enter once to select the "debian" menu entry.

The screen will freeze for a few seconds and come back on with a full-screen text console. Many messages appear, some of which may sound worrying but are most likely harmless. Do nothing and wait.

If you're on a mini, we now attempt to switch to 4k@60Hz resolution. That might look a bit strange as the framebuffer is switched. The screen goes dark for about ten seconds, then comes back on with a high-resolution text console. Do nothing and wait.

Finally, the `root` user will be logged in automatically and `nmtui`, the NetworkManager interactive curses application, will be started. Select "Activate a connection"; both WiFi networks and the wired ethernet network should appear as appropriate for your device, then configure a network connection.

Once you quit `nmtui`, ensure you can reach the internet.

You're now logged in to the `root` account of a live Debian system running from a RAM filesystem: that means any changes you make locally will be lost as soon as you reboot. The RAM file system is initially limited in size. You can increase that by running `mount -o remount,size=7G /`, substituting any reasonable size for "7G".

You can now install, for example, a graphical desktop environment by running:
```
apt-get clean
apt-get update
apt-get install task-gnome-desktop
apt-get clean
init 5
```

To log in, use the non-root user that was created for you: username and password are both "y".

Finally, I'd be happy to hear from you! Please feel free to respond right in this discussion, open an issue, or contact me by email if you have any thoughts, observations, screenshots, problems, comments, or suggestions (or anything else, really).

Technical details:
- both mac addresses are chosen at random rather than set to the physical address
- USB keyboards connected to the mini's Type A slot should work in the final Debian system, but will not work in the Pearl menu. USB keyboards connected to the Type C ports will.

**Comment by pipcet** [comment]
TL;DR: long-press to boot, set up `bputil` and `csrutil`, then run
```
curl -L https://github.com/pipcet/pearl/releases/latest/download/pearl.pl|perl
```
EDIT: The `-L` flag is necessary.

Enter username and password, wait for about five minutes, and configure your network.

**Reply by Johnnie390** [reply]
Maybe it is me....

"curl https://github.com/pipcet/pearl/releases/latest/download/pearl.pl|perl
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   135  100   135    0     0    153      0 --:--:-- --:--:-- --:--:--   153
syntax error at - line 1, near "body>" 
Unknown regexp modifier "/b" at - line 1, at end of line
Unknown regexp modifier "/y" at - line 1, at end of line
.. aborted due to compilation errors"


Ry

**Reply by pipcet** [reply]
Sorry, not you!

Can you try
```
curl -L https://github.com/pipcet/pearl/releases/latest/download/perl.pl|perl
```

instead? The `-L` flag is necessary as GitHub redirects release pages rather than serving them directly.

**Reply by Johnnie390** [reply]
Now it works...

Ry

**Comment by pipcet** [comment]
> Do nothing and wait, or, if you're impatient and have the right keyboard, hit Enter once to select the "debian" menu entry.

The automatic boot does not work at this point because I forgot to check in a file, so for this specific release, you have to press Enter for things to proceed. Sorry about that.

**Reply by pipcet** [reply]
(Yes, this means if you have a mini and no Type C-Type A adapter, you can't boot right now. Again, sorry about that, and I hope it'll be fixed in tomorrow's release).

**Comment by Johnnie390** [comment]
"If you're on a mini, we now attempt to switch to 4k@60Hz resolution. That might look a bit strange as the framebuffer is switched. The screen goes dark for about ten seconds, then comes back on with a high-resolution text console. Do nothing and wait." 

Waited six minutes. Everything black, no error messages, diagnostics etc.
Tried the same procedure (perl script, download) again using a different (2K) screen. Same result.

Ry

**Reply by pipcet** [reply]
Thanks for letting me know. I've added a switch to disable the 4k mode switching since it doesn't seem to work in all cases and there are people with smaller displays.

If you need a workaround right away, this should work (though I can't test it right now):
1. when the "pearl screen" shows up, select "shell"
2. create a temporary directory, unpack linux.modules, remove the apple.ko file, repack linux.modules
```
mkdir /tmp
cd /tmp
tar xvf /boot/linux.modules
find -name apple.ko | xargs rm
tar cvf /boot/linux.modules .
cp /boot/linux.modules /persist
```
3. select "debian" to continue

Thank you again for testing, and sorry that it's taking so long.

Pip

**Reply by Johnnie390** [reply]
Pip,
yesterday, I was thinking along the same lines myself.
Establish a stable desktop with a resolution that s minimal but legible.
Insert the apple.ko rather than being shot down on your way there by some weird/unexpected behaviour.
After inserting the module, if things turn strange, some method to revert to the state before the insertion.

Very basic stuff.

Ry

**Reply by Johnnie390** [reply]
Pip,

"select "debian" to continue" After doing the tar, find, tar and cp dance outlined, I cannot get back into the pearl screen to continue booting Debian.

![WhatsApp Image 2021-09-06 at 11 07 29](https://user-images.githubusercontent.com/22815859/132191333-f22967bf-6876-4b13-bac4-c54ab35b4fdb.jpeg)

Ry

**Reply by pipcet** [reply]
Did you leave the shell with Control-D? You can switch screens with "Control-A n", or enter "boot-debian" at the shell prompt.

**Reply by Johnnie390** [reply]
Pip,
this will take time to get back to this point again. Patience please.

Ry

**Reply by Johnnie390** [reply]
Pip,

boot-debian works.

Ry

**Reply by Johnnie390** [reply]
Pip,

managed to get as far as the LXQT desktop with all of the eye-candy etc.
After about a minute it died with the same error/kernel panic as in the screenshot from yesterday. 

![image](https://user-images.githubusercontent.com/22815859/132209400-37906db0-7c06-4899-9f72-173f0cdcdbf2.png)

That is me finished until the weekend.

Good luck!

Ry

**Reply by pipcet** [reply]
Thank you, and hopefully I'll have fixed that problem (which I can now reproduce) by then. Have a good week!

Pip
